api/utils/requestProcessor.js

```diff
@@ -2405,107 +2419,105 @@ const validateAppForFetchAPI = (params, done, try_times) => {
+                        common.db.collection('app_users' + params.app_id).update({
+                            _id: params.app_user_id,
+                            uid: {$exists: false}
+                        }, {$set: {uid: uid}}, {upsert: true, ignore_errors: [11000]}, function() {
+                            restartRequest(params, initiator, done, try_times, resolve);
+                        });
                     }
-                });
-            }
-            //check if device id was changed
-            else if (params.qstring.old_device_id && params.qstring.old_device_id !== params.qstring.device_id) {
-                const old_id = common.crypto.createHash('sha1')
-                    .update(params.qstring.app_key + params.qstring.old_device_id + "")
-                    .digest('hex');
-
-                countlyApi.mgmt.appUsers.merge(params.app_id, params.app_user, params.app_user_id, old_id, params.qstring.device_id, params.qstring.old_device_id, function() {
-                    //remove old device ID and retry request
-                    params.old_device_id = params.qstring.old_device_id; //Preserving old device id for logger
-                    params.qstring.old_device_id = null;
-                    restartFetchRequest(params, done, try_times, function() {
-                        return resolve();
-                    });
-                });
-            }
-            else {
-                return resolve();
-            }
-        });
+                }
+                else {
+                    //cannot create uid, so cannot process request now
+                    console.log("Cannot create uid", err, uid);
+                    resolve("Cannot create uid");
+                }
+            });
+        }
+        //check if device id was changed
+        else if (params.qstring.old_device_id && params.qstring.old_device_id !== params.qstring.device_id) {
+            const old_id = common.crypto.createHash('sha1')
+                .update(params.qstring.app_key + params.qstring.old_device_id + "")
+                .digest('hex');
+
+            countlyApi.mgmt.appUsers.merge(params.app_id, params.app_user, params.app_user_id, old_id, params.qstring.device_id, params.qstring.old_device_id, function(err) {
+                if (err) {
+                    return common.returnMessage(params, 400, 'Cannot update user');
+                }
+                //remove old device ID and retry request
+                params.qstring.old_device_id = null;
+                restartRequest(params, initiator, done, try_times, resolve);
+            });
+        }
+        else {
+            resolve();
+        }
     });
 }
+
 /**
  * @param  {object} params - params object
  * @param  {String} type - source type
  * @param  {Function} done - done callback
  * @returns {Function} - done or boolean value
  */
-const checksumSaltVerification = (params, type) => {
+const checksumSaltVerification = (params) => {
```

Comment:
Codacy found an issue: ['checksumSaltVerification' was used before it was defined.](https://app.codacy.com/manual/ar2rsawseen/countly-server/pullRequest?prid=6152164)